score.go

```diff
@@ -0,0 +1,885 @@
+	return nil
+}
+
+func (p *TopicScoreParams) validate() error {
+	// make sure we have a sane topic weight
+	if p.TopicWeight < 0 {
+		return fmt.Errorf("invalid topic weight; must be >= 0")
+	}
+
+	// check P1
+	if p.TimeInMeshWeight < 0 {
+		return fmt.Errorf("invalid TimeInMeshWeight; must be positive (or 0 to disable)")
+	}
+	if p.TimeInMeshWeight != 0 && p.TimeInMeshQuantum <= 0 {
+		return fmt.Errorf("invalid TimeInMeshQuantum; must be positive")
+	}
+	if p.TimeInMeshWeight != 0 && p.TimeInMeshCap <= 0 {
+		return fmt.Errorf("invalid TimeInMeshCap; must be positive")
+	}
+
+	// check P2
+	if p.FirstMessageDeliveriesWeight < 0 {
+		return fmt.Errorf("invallid FirstMessageDeliveriesWeight; must be positive (or 0 to disable)")
+	}
+	if p.FirstMessageDeliveriesWeight != 0 && (p.FirstMessageDeliveriesDecay <= 0 || p.FirstMessageDeliveriesDecay >= 1) {
+		return fmt.Errorf("invalid FirstMessageDeliveriesDecay; must be between 0 and 1")
+	}
+	if p.FirstMessageDeliveriesWeight != 0 && p.FirstMessageDeliveriesCap <= 0 {
+		return fmt.Errorf("invalid FirstMessageDeliveriesCap; must be positive")
+	}
+
+	// check P3
+	if p.MeshMessageDeliveriesWeight > 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesWeight; must be negative (or 0 to disable)")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && (p.MeshMessageDeliveriesDecay <= 0 || p.MeshMessageDeliveriesDecay >= 1) {
+		return fmt.Errorf("invalid MeshMessageDeliveriesDecay; must be between 0 and 1")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && p.MeshMessageDeliveriesCap <= 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesCap; must be positive")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && p.MeshMessageDeliveriesThreshold <= 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesThreshold; must be positive")
+	}
+	if p.MeshMessageDeliveriesWindow < 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesWindow; must be non-negative")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && p.MeshMessageDeliveriesActivation < time.Second {
+		return fmt.Errorf("invalid MeshMessageDeliveriesActivation; must be at least 1s")
+	}
+
+	// check P3b
+	if p.MeshFailurePenaltyWeight > 0 {
+		return fmt.Errorf("invalid MeshFailurePenaltyWeight; must be negative (or 0 to disable)")
+	}
+	if p.MeshFailurePenaltyWeight != 0 && (p.MeshFailurePenaltyDecay <= 0 || p.MeshFailurePenaltyDecay >= 1) {
+		return fmt.Errorf("invalid MeshFailurePenaltyDecay; must be between 0 and 1")
+	}
+
+	// check P4
+	if p.InvalidMessageDeliveriesWeight >= 0 {
+		return fmt.Errorf("invalid InvalidMessageDeliveriesWeight; must be negative")
+	}
+	if p.InvalidMessageDeliveriesDecay <= 0 || p.InvalidMessageDeliveriesDecay >= 1 {
+		return fmt.Errorf("invalid InvalidMessageDeliveriesDecay; must be between 0 and 1")
+	}
+
+	return nil
+}
+
+// router interface
+func (ps *peerScore) Start(gs *GossipSubRouter) {
+	ps.msgID = gs.p.msgID
+	ps.host = gs.p.host
+	go ps.background(gs.p.ctx)
+}
+
+func (ps *peerScore) Score(p peer.ID) float64 {
```

Comment:
As far as I can tell, this gets called for every single message received. It seems to me like it may get rather resource intensive. Let's keep this in mind and observe CPU profiles during tests

Reply:
Yes, the `Score` method gets called a lot. It should be relatively lightweight to compute, but good point about looking at CPU profiles.

Reply:
We also should watch lock contention there.